### ISSUE TITLE
ML curriculum expansion (Tier 1 + Tier 2): machine-learning/ + reinforcement-learning/

### ISSUE THREAD

**Issue by krisztiankoos** [kube-dojo/kube-dojo.github.io#677]
## Summary

KubeDojo's ML coverage has a real curriculum gap: only 3 modules in `classical-ml/` (sklearn survey, XGBoost, time-series), no `reinforcement-learning/` track anywhere, and several Tier-1 ML topics that any practitioner curriculum needs are missing entirely. Surfaced in handoff `docs/session-state/2026-04-30-night-7-followups.md` and locked across two cross-family consults (Codex + Gemini, 2026-05-01).

This issue tracks the ML expansion + reorganization work as a **standalone scope, separate from #388**. Different work type (NEW modules at full pedagogical depth, not regex-rubric satisfaction sweeps), different success criteria, different cadence.

## Locked structural plan

```
ai-ml-engineering/
├── machine-learning/        ← renamed from classical-ml/ (URLs preserved via redirects in astro.config.mjs)
│   └── 12 Tier-1 + 7 Tier-2 = 19 modules
├── reinforcement-learning/  ← NEW peer
│   └── 1 Tier-1 + 1 Tier-2 = 2 modules
└── deep-learning/           ← existing, +2 Tier-2 modules
```

`bridges/` stays as a peer (cross-track transition material). Other peers unchanged.

## Tier-1 — `machine-learning/` (12 modules, pedagogical order)

| # | Module | Status |
|---|---|---|
| 1.1 | Scikit-learn API & Pipelines | existing — needs **content rewrite** to drop algorithm survey overlap with 1.2 |
| 1.2 | Linear & Logistic Regression with Regularization | NEW |
| 1.3 | Model Evaluation, Validation, Leakage & Calibration | NEW |
| 1.4 | Feature Engineering & Preprocessing Systematic | NEW |
| 1.5 | Decision Trees & Random Forests | NEW |
| 1.6 | XGBoost & Gradient Boosting | existing |
| 1.7 | Naive Bayes, k-NN & SVMs | NEW |
| 1.8 | Unsupervised Learning: Clustering | NEW |
| 1.9 | Anomaly Detection & Novelty Detection | NEW |
| 1.10 | Dimensionality Reduction | NEW |
| 1.11 | Hyperparameter Optimization | NEW |
| 1.12 | Time Series Forecasting | existing |

## Tier-1 — `reinforcement-learning/` (1 module)

| # | Module | Status |
|---|---|---|
| 1.1 | RL Practitioner Foundations (PPO/DQN/SAC, SB3, Gymnasium, RLlib intro) | NEW |

## Tier-2 — `machine-learning/` (7 modules)

| # | Module |
|---|---|
| 2.1 | Class Imbalance & Cost-Sensitive Learning |
| 2.2 | ML Interpretability: SHAP, LIME, PDP/ICE + Failure Slicing & Counterfactual Debugging |
| 2.3 | Probabilistic & Bayesian ML with PyMC |
| 2.4 | Recommender Systems |
| 2.5 | Conformal Prediction & Uncertainty Quantification |
| 2.6 | Fairness & Bias Auditing |
| 2.7 | Causal Inference for ML Practitioners |

## Tier-2 — `reinforcement-learning/` (1 module)

| # | Module |
|---|---|
| 2.1 | Offline RL & Imitation Learning |

## Tier-2 — `deep-learning/` (2 modules, extending existing 7)

| # | Module |
|---|---|
| 1.8 | Self-Supervised Learning (SimCLR/MoCo, MAE, DINO) |
| 1.9 | Graph Neural Networks (PyTorch Geometric) |

## Source-grounding rails (mandatory per module)

Per `feedback_citation_verify_or_remove.md` and `feedback_gemini_hallucinates_anchors.md`, this expansion needs harder rails than prior content:

1. **Pre-vetted source list in brief**. Claude curates authoritative URLs (sklearn user guide, PyMC docs, MAPIE/Fairlearn/SB3 docs, named arxiv papers) before dispatch. Brief includes them explicitly.
2. **No-invention contract in dispatch prompt**. Codex forbidden from citing any URL not in the brief, inventing function signatures, fictional company anecdotes, or made-up benchmarks.
3. **Pre-merge citation script**: curl every URL in Sources section + grep cited claim. Fail on 404 / redirect / title mismatch. Enforced in code, not prompted.
4. **Code-block runnable check**: extract every ` ```python ` block, run import-and-syntax-check.
5. **Reviewer split**: Codex writes → Claude reviews teaching + factual claims → Gemini flags suspicious citations → Claude independently curl-verifies anything Gemini flags.

Citation-check + code-runnable scripts ship as part of Phase 0 so all subsequent modules inherit them.

## Phased execution

```
Phase 0 — Foundation PR (lands first, blocks Phase 1+)
  - mv classical-ml/ → machine-learning/
  - create reinforcement-learning/ scaffold
  - update astro.config.mjs sidebar + redirects
  - renumber existing 3 module slugs (XGBoost 1.2→1.6, time-series 1.3→1.12)
  - rewrite module 1.1 (refocus to API/pipelines)
  - ship citation-verify + code-runnable scripts under scripts/quality/
  - update changelog.md, parent index.md tables
  - Codex cross-family review
  - merge

Phase 1a — Tier-1 ML cross-cutting (used by everything else)
  - 1.3 Evaluation/Validation/Leakage/Calibration
  - 1.4 Feature Engineering & Preprocessing

Phase 1b — Tier-1 ML algorithms (pedagogical order)
  - 1.2, 1.5, 1.7, 1.8, 1.9, 1.10, 1.11

Phase 2 — RL T1 + DL T2
  - reinforcement-learning/1.1 RL Practitioner Foundations
  - deep-learning/1.8 Self-Supervised Learning
  - deep-learning/1.9 Graph Neural Networks

Phase 3 — Tier-2 ML completeness
  - 2.1 through 2.7
  - reinforcement-learning/2.1 Offline RL & Imitation Learning
```

## Per-module workflow (Phase 1+)

Each module follows the same loop:

1. Brief contract written (Claude, ~200-400 words): audience, prereqs, Bloom L3+ outcomes, line target 600-800+, key concepts, pitfalls, pre-vetted source URLs.
2. Codex draft (gpt-5.5, model_reasoning_effort=high, workspace-write + per-issue worktree).
3. Claude review for accuracy + teaching quality (not listicle per `feedback_teaching_not_listicles.md`).
4. Gemini citation flagging (--review).
5. Claude independent curl-verifies any flagged URL.
6. Quality gates: build green, health check 0 errors, new-content-checklist (frontmatter, slug, parent index, internal links, changelog), citation-check script, code-runnable script.
7. Cross-family review per `docs/review-protocol.md` (Codex-authored → Claude reviewer).
8. Branch + PR + merge.

## Constraints

- Sequential Codex dispatches only (per `feedback_codex_dispatch_sequential.md`).
- No parallel Agent worktree fan-out (per `feedback_no_parallel_agent_fanout.md`).
- One PR per module (except Phase 0 which bundles foundation).
- Cross-link, don't duplicate: drift detection stays in `mlops/module-1.10-ml-monitoring.md`; RLHF stays in `advanced-genai/`.

## Refs

- Handoff: `docs/session-state/2026-04-30-night-7-followups.md`
- Sister scope: #388 (existing module rewrite — does not double-count this work)
- Quality bar: `docs/quality-rubric.md` ≥33/40 sum AND every dimension ≥4
- Module standards: `.claude/rules/module-quality.md`
- Content checklist: `.claude/rules/new-content-checklist.md`
- Review protocol: `docs/review-protocol.md`

## Cross-thread note for #388

Modules in `machine-learning/` and `reinforcement-learning/` are being authored at full pedagogical depth under this issue. Do not double-count them in #388's critical-rubric sweeps. New modules will satisfy the regex rubric (Sources / quiz / exercise / diagram) by construction via Phase 0's quality gates.

**Comment by krisztiankoos** [comment]
## Closing — ML/RL/DL curriculum expansion complete

All 22 work items shipped (19 from the 2026-05-01 batch + 3 from this session: ML 2.7 PR #708, RL 2.1 PR #709, DL 1.7 cleanup PR #710).

### Tier 1 — Machine Learning (12 modules)
- 1.1 Scikit-Learn API & Pipelines (Phase 0 rewrite, PR #680)
- 1.2 Linear & Logistic Regression with Regularization (PR #692)
- 1.3 Model Evaluation, Validation, Leakage & Calibration (PR #686)
- 1.4 Feature Engineering & Preprocessing (PR #691)
- 1.5 Decision Trees & Random Forests (PR #693)
- 1.7 Naive Bayes, k-NN & SVMs (PR #694)
- 1.8 Unsupervised Learning: Clustering (PR #695)
- 1.9 Anomaly Detection & Novelty Detection (PR #696)
- 1.10 Dimensionality Reduction (PR #697)
- 1.11 Hyperparameter Optimization (PR #698)

### Tier 2 — Machine Learning (7 modules)
- 2.1 Class Imbalance & Cost-Sensitive Learning (PR #702)
- 2.2 Interpretability + Failure Slicing (PR #703)
- 2.3 Probabilistic & Bayesian ML with PyMC (PR #704)
- 2.4 Recommender Systems (PR #705)
- 2.5 Conformal Prediction & Uncertainty Quantification (PR #706)
- 2.6 Fairness & Bias Auditing (PR #707)
- 2.7 Causal Inference for ML Practitioners (PR #708)

### Reinforcement Learning (2 modules — full track)
- 1.1 RL Practitioner Foundations (PR #699)
- 2.1 Offline RL & Imitation Learning (PR #709)

### Deep Learning (2 new + 1 fix)
- 1.8 Self-Supervised Learning (PR #700)
- 1.9 Graph Neural Networks (PR #701)
- 1.7 filename/slug/forward-link cleanup (PR #710)

### Operational lessons captured
- Codex authored 14 of 20 modules; orchestrator Claude wrote 5 fallback modules + Phase 0 + reviewed all Codex-authored. Cross-family review caught 6 distinct teaching bugs that orchestrator self-review would have missed.
- Pinned-slug discipline + plain-text Next-Module convention + no-skill / no-cold-start preamble eliminated the dispatch hallucination class.
- Codex caught 4 external-source bugs pre-dispatch: `arxiv:1212.5745` was dark-matter physics (not Bergstra & Bengio), `arxiv:1610.04018` was colloidal physics (not Athey & Wager), `arxiv:1706.03461` had been reused for two different papers in a brief, `implicit.readthedocs.io` is hijacked spam, MAPIE v1 API is breaking-change, ArviZ moved to `python.arviz.org`. All findings memorialized in agent memory + cross-thread notes.

Track headcounts after this issue: machine-learning 19 modules, reinforcement-learning 2 modules, deep-learning 9 modules.
